### PR TITLE
[SPARK-22856][SQL] Add wrappers for codegen output and nullability

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

The codegen output of `Expression`, aka `ExprCode`, now encapsulates only strings of output value (`value`) and nullability (`isNull`). It makes difficulty for us to know what the output really is. I think it is better if we can add wrappers for the value and nullability that let us to easily know that.

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @kiszk @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -56,7 +56,36 @@ import org.apache.spark.util.{ParentClassLoader, Utils}
...
+  override def toString: String = variableName
+}
+
+// A statement evaluation of [[ExprCode]].
+case class StatementValue(val statement: String) extends ExprValue {
+  override def toString: String = statement
+}
+
+// A global variable evaluation of [[ExprCode]].
+case class GlobalValue(val value: String) extends ExprValue {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: for compacted global variables, we may get something like `arr[1]` while `arr` is a global variable. Is `arr[1]` a statement or global variable?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: It is considered as global variable now, as it can be accessed globally and don't/can't be parameterized.

**Comment by viirya** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
Oh, already re-testing.

**Comment by viirya** [comment]
Thanks @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala [hunk 2]
```diff
@@ -321,7 +322,12 @@ case class IsNull(child: Expression) extends UnaryExpression with Predicate {
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val eval = child.genCode(ctx)
-    ExprCode(code = eval.code, isNull = "false", value = eval.isNull)
+    val value = if ("true" == s"${eval.isNull}" || "false" == s"${eval.isNull}") {
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: `if ("true" == eval.isNull || "false" == eval.isNull) {`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala [hunk 3]
```diff
@@ -347,7 +353,14 @@ case class IsNotNull(child: Expression) extends UnaryExpression with Predicate {
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val eval = child.genCode(ctx)
-    ExprCode(code = eval.code, isNull = "false", value = s"(!(${eval.isNull}))")
+    val value = if ("true" == s"${eval.isNull}") {
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 4]
```diff
@@ -118,10 +118,10 @@ abstract class Expression extends TreeNode[Expression] {
   private def reduceCodeSize(ctx: CodegenContext, eval: ExprCode): Unit = {
     // TODO: support whole stage codegen too
     if (eval.code.trim.length > 1024 && ctx.INPUT_ROW != null && ctx.currentVars == null) {
-      val setIsNull = if (eval.isNull != "false" && eval.isNull != "true") {
+      val setIsNull = if ("false" != s"${eval.isNull}" && "true" != s"${eval.isNull}") {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: this can be simplified to `!eval.isNull.instanceOf[LiteralValue]`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 5]
```diff
@@ -56,7 +56,36 @@ import org.apache.spark.util.{ParentClassLoader, Utils}
  * @param value A term for a (possibly primitive) value of the result of the evaluation. Not
  *              valid if `isNull` is set to `true`.
  */
-case class ExprCode(var code: String, var isNull: String, var value: String)
+case class ExprCode(var code: String, var isNull: ExprValue, var value: ExprValue)
+
+
+// An abstraction that represents the evaluation result of [[ExprCode]].
+abstract class ExprValue
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I think we should classify `ExprValue` by our needs, not by java definitions. Thinking about the needs, we wanna know: 1) if this value is accessible anywhere and we don't need to carry it via method parameters. 2) if this value needs to be carried with parameters, do we need to generate a parameter name or use this value directly?

So basically we can combine `LiteralValue` and `GlobalValue`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala (hunk 2 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: or `eval.isNull == Literal("true")`? Or even better we can create a `LiteralTrue = Literal("true")` and equivalent for false and use them?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 5 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: IMHO I prefer this approach because in the future we might need to distinguish these two cases, thus I think is a good thing to let them be distinct.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: oh, yea.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 5 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: For now `LiteralValue` and `GlobalValue` can be seen as the same effectively, as they are all accessible anywhere and we don't need to carry it via method parameters.

I don't have strong preference here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala]: We can do `eval.isNull.instanceOf[LiteralValue]` as suggested by @cloud-fan below.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 5 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: @kiszk WDYT?

Comment (kiszk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: In summary, I have no strong preference.

In the future, we will want to distinguish `Literal` and `Global` for some optimizations. [This](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala#L121) is already one of optimizations for `Literal`.

If this PR just focuses on classifying types between arguments and non-arguments, it is fine to combine `Literal` and `Global`. Then, another PR will separate one type into `Literal` and `Global`.

**Comment by viirya** [comment]
Thanks.

If we have forseen this need in near future, do we bother to combine them
here and separate it again later?

On Dec 22, 2017 12:02 PM, "Kazuaki Ishizaki" <notifications@github.com>
wrote:

*@kiszk* commented on this pull request.
------------------------------

In sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/
expressions/codegen/CodeGenerator.scala
<https://github.com/apache/spark/pull/20043#discussion_r158426351>:

> @@ -56,7 +56,36 @@ import org.apache.spark.util.{ParentClassLoader, Utils}
  * @param value A term for a (possibly primitive) value of the result
of the evaluation. Not
  *              valid if `isNull` is set to `true`.
  */
-case class ExprCode(var code: String, var isNull: String, var value: String)
+case class ExprCode(var code: String, var isNull: ExprValue, var
value: ExprValue)
+
+
+// An abstraction that represents the evaluation result of [[ExprCode]].
+abstract class ExprValue

In summary, I have no strong preference.

In the future, we will want to distinguish Literal and Global for some
optimizations. This
<https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala#L121>
is already one of optimizations for Literal.

If this PR just focuses on classifying types between arguments and
non-arguments, it is fine to combine Literal and Global. Then, another PR
will separate one type into Literal and Global.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/apache/spark/pull/20043#discussion_r158426351>, or mute
the thread

**Comment by gczsjdy** [comment]
@viirya Sorry I didn't quite understand, how do we easily know the value by adding wrappers? Could you explain a little bit?

**Comment by viirya** [comment]
@gczsjdy By this, we can easily know what the output is. Is it a global variable, a local variable or a literal? For now, we just get a string. It makes us hard to act according to the type of codegen output. It's also hard to search the codes to find out where we output global variable, local variable or literal, BTW.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 5 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: If no strong preference for combining them, I'd keep it as two concepts for now, if we foresee the need to distinguish them.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: @cloud-fan What do you think?

**Comment by gczsjdy** [comment]
@viirya Thanks much. Actually local variable corresponds to `VariableValue` and `StatementValue`? IIUC `VariableValue` is value that depends on something else, but what is `StatementValue`? Maybe we can add more comments near the `xxValue` definition.

**Comment by viirya** [comment]
`StatementValue` means an output like `a + 1`. It is a java statement which doesn't rely on a local variable to hold the result.

**Comment by viirya** [comment]
ping @cloud-fan

**Comment by viirya** [comment]
@cloud-fan Any more comments on this?

**Comment by viirya** [comment]
ping @cloud-fan

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
@cloud-fan Can you take a look again? Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: OK let's keep it.

**Comment by cloud-fan** [comment]
Sorry I think we can't make this into Spark 2.3. Recently I'm busy with some data source v2 features that are promised for 2.3, I will come back to this after rc1.

**Comment by viirya** [comment]
Sure. No problem.

**Comment by rednaxelafx** [comment]
Hi @viirya, this PR is definitely onto the right direction. I'd like to review this PR as well and hopeful help polish it in for post-2.3. Thanks for your work!

**Comment by viirya** [comment]
@rednaxelafx Thanks! Yeah, let's revisit this after 2.3.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala [hunk 6]
```diff
@@ -75,7 +75,7 @@ case class BoundReference(ordinal: Int, dataType: DataType, nullable: Boolean)
              |$javaType ${ev.value} = ${ev.isNull} ? ${ctx.defaultValue(dataType)} : ($value);
            """.stripMargin)
       } else {
-        ev.copy(code = s"$javaType ${ev.value} = $value;", isNull = "false")
+        ev.copy(code = s"$javaType ${ev.value} = $value;", isNull = LiteralValue("false"))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala]: nit: shall we introduce a `TrueLiteral` and `FalseLiteral`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala]: I think it should be useful to the `isNull` field.

**Comment by cloud-fan** [comment]
Spark 2.3 should be ready soon, let's resume this PR :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala (hunk 6 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala]: Looks like a good idea.

**Comment by viirya** [comment]
Just resolve conflicts and add `TrueLiteral` and `FalseLiteral` first. Will replace literals with `TrueLiteral` and `FalseLiteral` later.

**Comment by viirya** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 7]
```diff
@@ -323,7 +323,8 @@ class CodegenContext {
       case _: StructType | _: ArrayType | _: MapType => s"$value = $initCode.copy();"
       case _ => s"$value = $initCode;"
     }
-    ExprCode(code, "false", value)
+    ExprCode(code, FalseLiteral,
+      GlobalValue(value, ExprType(this, dataType)))
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: nit: this can go on one line

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+case class GlobalValue(val value: String, val javaType: ExprType) extends ExprValue {
+  override def toString: String = value
+  override val canDirectAccess: Boolean = true
+}
+
+case object TrueLiteral extends LiteralValue("true", ExprType("boolean", true))
+case object FalseLiteral extends LiteralValue("false", ExprType("boolean", true))
+
+// Represents the java type of an evaluation.
+case class ExprType(val typeName: String, val isPrimitive: Boolean)
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: why is this `isPrimitive` needed? If I am not wrong, we have somewhere a method to check whether a type is primitive or not. I think we can get rid of this and use that method when needed, or at least store this using that method instead of passing it every time.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 9]
```diff
@@ -31,7 +31,7 @@ import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.ScalaReflection.universe.TermName
 import org.apache.spark.sql.catalyst.encoders.RowEncoder
 import org.apache.spark.sql.catalyst.expressions._
-import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
+import org.apache.spark.sql.catalyst.expressions.codegen._
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: can we list the needed classes instead?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  def apply(value: String, javaType: ExprType): LiteralValue = new LiteralValue(value, javaType)
+  def unapply(literal: LiteralValue): Option[(String, ExprType)] =
+    Some((literal.value, literal.javaType))
+}
+
+// A variable evaluation of [[ExprCode]].
+case class VariableValue(
+    val variableName: String,
+    val javaType: ExprType,
+    val canDirectAccess: Boolean = false) extends ExprValue {
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: why isn't this fixed like for `GlobalValue`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 11]
```diff
@@ -22,7 +22,7 @@ import scala.collection.mutable.ArrayBuffer
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions._
-import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
+import org.apache.spark.sql.catalyst.expressions.codegen._
```
Comment (mgaido91) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala (hunk 8 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: Here the idea is to include java type information for an evaluation. Then we don't need to consult `CodegenContext`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 9 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: It will list too many classes `CodegenContext`, `ExprCode`, `ExprValue`, `GlobalValue`, `FalseLiteral`, `VariableValue`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala (hunk 8 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: than can't we move the method from `CodegenContext` to a static method and use that? Currently this information is never used, and I feel this is hard to maintain (even though I don't expect it to change frequently).

We can add a method like
```
def isPrimitive: Boolean = CodegenContext.isPrimitive(typeName)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala (hunk 10 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: a static variable is a `GlobalValue`, isn't it? Considering that we should be able to access also from methods in other internal classes I don't see any use case where this flexibility is required, honestly...

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: Ok. I'd let it as fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala [hunk 12]
```diff
@@ -70,13 +72,14 @@ case class GlobalValue(val value: String, val javaType: ExprType) extends ExprVa
...
 // Represents the java type of an evaluation.
-case class ExprType(val typeName: String, val isPrimitive: Boolean)
+case class ExprType(val typeName: String) {
+  def isPrimitive(ctx: CodegenContext): Boolean = ctx.isPrimitiveType(typeName)
+}
 
 object ExprType {
-  def apply(ctx: CodegenContext, dataType: DataType): ExprType = ExprType(ctx.javaType(dataType),
-    ctx.isPrimitiveType(dataType))
+  def apply(ctx: CodegenContext, dataType: DataType): ExprType = ExprType(ctx.javaType(dataType))
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: I am not sure this is useful. `ctx.javaType(dataType)` can be done by the caller

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala [hunk 13]
```diff
@@ -70,13 +72,14 @@ case class GlobalValue(val value: String, val javaType: ExprType) extends ExprVa
...
 
-case object TrueLiteral extends LiteralValue("true", ExprType("boolean", true))
-case object FalseLiteral extends LiteralValue("false", ExprType("boolean", true))
+case object TrueLiteral extends LiteralValue("true", ExprType("boolean"))
+case object FalseLiteral extends LiteralValue("false", ExprType("boolean"))
 
 // Represents the java type of an evaluation.
-case class ExprType(val typeName: String, val isPrimitive: Boolean)
+case class ExprType(val typeName: String) {
+  def isPrimitive(ctx: CodegenContext): Boolean = ctx.isPrimitiveType(typeName)
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: I see that @kiszk created a PR to move all the "static" methods to `CodeGenerator`. Using that approach here we do not need to pass `ctx` as an argument.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+  override def toString: String = value
+  override val canDirectAccess: Boolean = true
+}
+
+case object TrueLiteral extends LiteralValue("true", ExprType("boolean"))
+case object FalseLiteral extends LiteralValue("false", ExprType("boolean"))
+
+// Represents the java type of an evaluation.
+case class ExprType(val typeName: String) {
+  def isPrimitive(ctx: CodegenContext): Boolean = ctx.isPrimitiveType(typeName)
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: here following the approach in #20700 we can avoid to pass `CodegenContext` as a parameter. We can move the method to `CodeGenerator`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+object ExprType {
+  def apply(ctx: CodegenContext, dataType: DataType): ExprType = ExprType(ctx.javaType(dataType))
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: I don't think this method is very useful. If needed, the caller can do `ctx.javaType(dataType)` I think

**Comment by mgaido91** [comment]
this LGTM, any more comments @cloud-fan @kiszk @rednaxelafx ?

**Comment by mgaido91** [comment]
retest this please

**Comment by hvanhovell** [comment]
@viirya big fan of this change! More structure will make code gen easier & safer to implement. I think we should merge this as is, and then I think it might be good to start adding types to the values, and to make the `CodeGenerator` and the `CodegenContext` work directly with these values.

Since I merged @kiszk PR just now, can you update? I am sorry for the hassle.

**Comment by viirya** [comment]
@hvanhovell Thanks! I also think this structure can help us improve codegen. I will update it soon.

**Comment by viirya** [comment]
ping @hvanhovell @cloud-fan Any more comment for this?

**Comment by kiszk** [comment]
LGTM

**Comment by viirya** [comment]
ping @hvanhovell @cloud-fan

**Comment by viirya** [comment]
retest this please.

**Comment by hvanhovell** [comment]
retest this please

**Comment by hvanhovell** [comment]
I am going to merge this after this successfully passes tests

**Comment by viirya** [comment]
retest this please.

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by viirya** [comment]
Thanks! @cloud-fan @hvanhovell @kiszk @HyukjinKwon @mgaido91 @maropu @rednaxelafx @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+  def isPrimitive: Boolean = CodeGenerator.isPrimitiveType(javaType)
+}
+
+object ExprValue {
+  implicit def exprValueToString(exprValue: ExprValue): String = exprValue.toString
+}
+
+// A literal evaluation of [[ExprCode]].
+class LiteralValue(val value: String, val javaType: String) extends ExprValue {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: why not a case class?

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: We currently have case objects for `TrueLiteral` and `FalseLiteral` which extends `LiteralValue`.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExprValue.scala]: Yes.
